pysmu.cpp

```diff
@@ -3,267 +3,425 @@
+		if (ret != 0)
+			Py_RETURN_FALSE;
+		Py_RETURN_TRUE;
+	}
+
+	static PyObject* cleanupSession(PyObject* self, PyObject* args){
+		session->end();
+		return PyInt_FromLong(0l);
+	}
+
+	static PyObject* getDevInfo(PyObject* self, PyObject* args){
+		PyObject* data = PyList_New(0);
+		for (auto i: session->m_available_devices){
+			session->add_device(&*i);
+		}
+		for (auto dev: session->m_devices){
+			auto dev_info = dev->info();
+			PyObject* dev_data = PyDict_New();
+			for (unsigned chan=0; chan < dev_info->channel_count; chan++) {
+				auto chan_info = dev->channel_info(chan);
+				dev->set_mode(chan, 1);
+				PyObject* sigs = PyList_New(0);
+
+				for (unsigned sig=0; sig < chan_info->signal_count; sig++) {
+					auto info = dev->signal(chan, sig)->info()->label;
+					PyObject* info_str = PyString_FromString(string(info).c_str());
+					PyList_Append(sigs, info_str);
+					Py_DECREF(info_str);
+				}
+				PyObject* chan_label = PyString_FromString(string(chan_info->label).c_str());
+				PyDict_SetItem(dev_data, chan_label, sigs);
+				Py_DECREF(chan_label);
+				Py_DECREF(sigs);
+			}
+			PyObject* serial_str = PyString_FromString(dev->serial());
+			PyObject* ret = PyTuple_Pack(2, serial_str, dev_data);
+			Py_DECREF(serial_str);
+			Py_DECREF(dev_data);
+			PyList_Append(data, ret);
+			Py_DECREF(ret);
+		}
+		return data;
+	}
+
+	static shared_ptr<Device> get_device(int dev_num) {
+		if (dev_num < 0) {
+			PyErr_SetString(PyExc_ValueError, "invalid device ID");
+			return NULL;
+		}
+		auto dev = session->get_device((unsigned) dev_num);
+		if (dev == NULL) {
+			PyErr_SetString(PyExc_ValueError, "device not found");
+			return NULL;
+		}
+		return dev;
+	}
+
+	static PyObject* setMode(PyObject* self, PyObject* args){
+		int dev_num;
+		int chan_num;
+		int mode_num;
+		if (!PyArg_ParseTuple(args, "iii", &dev_num, &chan_num, &mode_num))
+			return NULL;
+
+		auto dev = get_device(dev_num);
+		if (dev == NULL)
+			return NULL;
+		dev->set_mode((unsigned) chan_num, (unsigned) mode_num);
+		Py_RETURN_NONE;
+	}
+
+	static PyObject* getInputs(PyObject* self, PyObject* args){
+		int dev_num;
+		int chan_num;
+		int nsamples;
+		if (!PyArg_ParseTuple(args, "iii", &dev_num, &chan_num, &nsamples))
+			return NULL;
+
+		auto dev = get_device(dev_num);
+		if (dev == NULL)
+			return NULL;
+		auto sgnl_v = dev->signal(chan_num, 0);
+		auto sgnl_i = dev->signal(chan_num, 1);
+		vector<float> buf_v;
+		vector<float> buf_i;
+		buf_v.resize(nsamples);
+		buf_i.resize(nsamples);
+		sgnl_v->measure_buffer(buf_v.data(), nsamples);
+		sgnl_i->measure_buffer(buf_i.data(), nsamples);
+		session->configure(SAMPLE_RATE);
+		session->run(nsamples);
+		PyObject* samples = PyList_New(0);
+		for (int i = 0; i < nsamples; i++) {
+			PyObject* sample_tuple = Py_BuildValue("(f,f)", buf_v[i], buf_i[i]);
+			PyList_Append(samples, sample_tuple);
+			Py_DECREF(sample_tuple);
+		}
+		return samples;
+	}
+
+	static PyObject* getAllInputs(PyObject* self, PyObject* args) {
```

Comment:
@radhermit it looks like this method only returns all inputs across a given device?